src/manager.cpp

```diff
@@ -0,0 +1,225 @@
+    try
+    {
+        auto numSessId = parseSessionId(sessionId);
+        if (sessionItems.find(numSessId) == sessionItems.end())
+        {
+            throw InvalidArgument();
+        }
+        sessionItems.erase(numSessId);
+    }
+    catch (const std::exception& e)
+    {
+        log<level::ERR>("Failure to close an obmc session.",
+                        entry("SESSION-ID=%s", sessionId.c_str()),
+                        entry("ERROR=%s", e.what()));
+        throw InternalFailure();
+    }
+}
+
+std::size_t SessionManager::removeAll(const std::string& userName)
+{
+    const auto count =
+        std::erase_if(sessionItems, [userName](const auto& item) {
+            auto const [id, session] = item;
+            return session->getOwner() == userName;
+        });
+
+    return count;
+}
+
+std::size_t
+    SessionManager::removeAllByRemoteAddress(const std::string& remoteAddress)
+{
+    const auto count =
+        std::erase_if(sessionItems, [remoteAddress](const auto& item) {
+            auto const [id, session] = item;
+            return session->remoteIPAddr() == remoteAddress;
+        });
+
+    return count;
+}
+
+std::size_t SessionManager::removeAll()
+{
+    size_t handledSessions = sessionItems.size();
+    sessionItems.clear();
+    return handledSessions;
+}
+
+SessionIdentifier SessionManager::generateSessionId() const
+{
+    auto time = std::chrono::high_resolution_clock::now();
+    std::size_t timeHash =
+        std::hash<int64_t>{}(time.time_since_epoch().count());
+    std::size_t serviceNameHash = std::hash<std::string>{}(serviceName);
+
+    // constexpr unsigned int leftShiftSize = 8 * sizeof(std::size_t);
+    std::size_t result = timeHash ^ (serviceNameHash << 1);
+    if (result == invalidSessionId)
+    {
+        // The hash-collision guard. The Session ID == invalidSessionId is reserved and can't
+        // be provided as a valid ID.
+        return generateSessionId();
+    }
+
+    return result;
+}
+
+const std::string
+    SessionManager::getSessionObjectPath(SessionIdentifier sessionId) const
+{
+    return getSessionManagerObjectPath() + "/" + hexSessionId(sessionId);
+}
+
+const std::string SessionManager::getSessionManagerObjectPath() const
+{
+    return std::string(sessionManagerObjectPath);
+}
+
+const std::string SessionManager::hexSessionId(SessionIdentifier sessionId)
+{
+    std::stringstream stream;
+    stream << std::setfill('0') << std::setw(sizeof(sessionId) * 2) << std::hex
+           << sessionId;
+    return stream.str();
+}
+
+SessionIdentifier SessionManager::parseSessionId(const std::string hexSessionId)
+{
+    return std::stoull(hexSessionId, nullptr, 16);
+}
+
+void SessionManager::checkSessionOwnerAlive(const boost::system::error_code&)
+{
+    bool loopChecker;
+    // Do loop through 'while' upon each found unreachable session because after
+    // closing session the SessionItems iterator will be invalidated.
```

Comment:
Special for the such cases `std::map::erase(iterator)` returns an iterator pointing at the next element.
So, I guess this may be rewritten to something like:
```cpp
for (auto it = sessionItems.begin(); it != sessionItems.end();)
{
    if (!fs::exists(it->second->getProcPath()))
    {
        ...
        it = sessionItems.erase(it);
    }
    else
    {
        ++it;
    }
}
```


Your implementation should also work properly, but a little bit slower because it will start a new loop through all valid elements after each removal.

Reply:
Yes, at the worst, my implementation will give O(fact n), but your suggested idiom provides O(log n), and that is too much better.

Reply:
Fixed